gst/mqtt/mqttsrc.c

```diff
@@ -0,0 +1,842 @@
+}
+
+/**
+ * @brief Setter for the 'cleansession' property.
+ */
+static void
+gst_mqtt_src_set_opt_cleansession (GstMqttSrc * self, const gboolean val)
+{
+  self->mqtt_conn_opts.cleansession = val;
+}
+
+/**
+ * @brief Getter for the 'keep-alive-interval' property
+ */
+static gint
+gst_mqtt_src_get_opt_keep_alive_interval (GstMqttSrc * self)
+{
+  return self->mqtt_conn_opts.keepAliveInterval;
+}
+
+/**
+ * @brief Setter for the 'keep-alive-interval' property
+ */
+static void
+gst_mqtt_src_set_opt_keep_alive_interval (GstMqttSrc * self, const gint num)
+{
+  self->mqtt_conn_opts.keepAliveInterval = num;
+}
+
+/**
+  * @brief A callback to handle the connection lost to the broker
+  */
+static void
+cb_mqtt_on_connection_lost (void *context, char *cause)
+{
+  GstMqttSrc *self = GST_MQTT_SRC_CAST (context);
+
+  g_mutex_lock (&self->mqtt_src_mutex);
+  self->is_connected = FALSE;
+  self->is_subscribed = FALSE;
+  g_cond_broadcast (&self->mqtt_src_gcond);
+  self->err = g_error_new (self->gquark_err_tag, EHOSTDOWN,
+      "Connection to the host (broker) has been lost: %s",
+      g_strerror (EHOSTDOWN));
+  g_mutex_unlock (&self->mqtt_src_mutex);
+}
+
+/**
+  * @brief A callback to handle the post-processing of the delivered message
+  * @todo Fill the function body
+  */
+static void
+cb_mqtt_on_delivery_complete (void *context, MQTTAsync_token token)
+{
+
+}
+
+/**
+  * @brief A callback to handle the arrived message
+  */
+static int
+cb_mqtt_on_message_arrived (void *context, char *topic_name, int topic_len,
+    MQTTAsync_message * message)
+{
+  const int size = message->payloadlen;
+  guint8 *data = message->payload;
+  GstMQTTMessageHdr *mqtt_msg_hdr;
+  GstMapInfo hdr_map_info;
+  GstMemory *recieved_mem;
+  GstMemory *hdr_mem;
+  GstBuffer *buffer;
+  GstMqttSrc *self;
+  gsize offset;
+  guint i;
+
+  self = GST_MQTT_SRC_CAST (context);
+  recieved_mem = gst_memory_new_wrapped (0, data, size, 0, size, message,
+      (GDestroyNotify) cb_memory_wrapped_destroy);
+  if (!recieved_mem) {
+    self->err = g_error_new (self->gquark_err_tag, ENODATA,
+        "%s: failed to wrap the raw data of recieved message in GstMemory: %s",
+        __func__, g_strerror (ENODATA));
+    return TRUE;
+  }
+
+  mqtt_msg_hdr = _extract_mqtt_msg_hdr_from (recieved_mem, &hdr_mem,
+      &hdr_map_info);
+  if (!mqtt_msg_hdr) {
+    self->err = g_error_new (self->gquark_err_tag, ENODATA,
+        "%s: failed to extract header information from recieved message: %s",
+        __func__, g_strerror (ENODATA));
+    goto ret_unref_recieved_mem;
+  }
+
+  buffer = gst_buffer_new ();
+  offset = GST_MQTT_LEN_MSG_HDR;
+  for (i = 0; i < mqtt_msg_hdr->num_mems; ++i) {
+    GstMemory *each_memory;
+    int each_size;
+
+    each_size = mqtt_msg_hdr->size_mems[i];
+    each_memory = gst_memory_share (recieved_mem, offset, each_size);
+    gst_buffer_append_memory (buffer, each_memory);
+    offset += each_size;
+  }
+
+  g_async_queue_push (self->aqueue, buffer);
```

Comment:
At this point, if the buffer is empty, which decision is right?
1. send empty buffer
2. return false